File: frontend/src/api/skills.ts
const apiBase = import.meta.env.VITE_API_URL ?? ''

export type SkillOut = {
  id: string
  title: string
  notes: string | null
  context: Record<string, unknown> | null
  stats_sessions: number
  stats_practice_seconds: number
  stats_level: number
  stats_progress_percent: number
  stats_mastered: number
  stats_day_streak: number
  last_practice_at: string | null
  created_at: string
  updated_at: string
}

export type ResearchOut = {
  id: string
  skill_id: string
  title: string | null
  content: string
  extra: Record<string, unknown> | null
  created_at: string
}

export type SkillWithResearchResponse = {
  skill: SkillOut
  research: ResearchOut
}

export type SkillCreateWithResearchBody = {
  title: string
  goal: string
  level: string
  category?: string | null
}

export type SessionCompleteBody = {
  duration_seconds: number
  session_notes?: string | null
}

export type SkillSessionSummaryOut = {
  id: string
  skill_id: string
  session_number: number
  duration_seconds: number
  summary_text: string
  coach_note: string | null
  progress_delta: number
  level_ups: number
  mastered_delta: number
  input_notes: string | null
  extra: Record<string, unknown> | null
  created_at: string
}

export type SessionCompleteResponse = {
  skill: SkillOut
  coach_note: string
  progress_delta: number
  level_ups: number
  mastered_delta: number
  session_summary: SkillSessionSummaryOut
  docs_export_url: string | null
}

export type ProgressEventOut = {
  id: string
  skill_id: string
  kind: string
  label: string | null
  detail: Record<string, unknown> | null
  metric_value: number | null
  created_at: string
}

export async function fetchSkills(): Promise<SkillOut[]> {
  const res = await fetch(`${apiBase}/api/skills`, { credentials: 'include' })
  if (res.status === 401) {
    throw new Error('Sign in to load skills.')
  }
  if (!res.ok) {
    throw new Error(`Failed to load skills: ${res.status}`)
  }
  const data = (await res.json()) as { skills: SkillOut[] }
  return data.skills
}

export async function fetchSkill(skillId: string): Promise<SkillOut> {
  const res = await fetch(`${apiBase}/api/skills/${encodeURIComponent(skillId)}`, {
    credentials: 'include',
  })
  if (res.status === 401) {
    throw new Error('Sign in to load this skill.')
  }
  if (!res.ok) {
    throw new Error(`Failed to load skill: ${res.status}`)
  }
  return res.json() as Promise<SkillOut>
}

function errorFromResponse(res: Response, bodyText: string): string {
  try {
    const data = JSON.parse(bodyText) as { detail?: unknown }
    if (typeof data.detail === 'string') return data.detail
  } catch {
    /* ignore */
  }
  return bodyText || `Request failed: ${res.status}`
}

export async function createSkillWithResearch(
  body: SkillCreateWithResearchBody,
): Promise<SkillWithResearchResponse> {
  const res = await fetch(`${apiBase}/api/skills/create-with-research`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const text = await res.text()
  if (!res.ok) {
    throw new Error(errorFromResponse(res, text))
  }
  return JSON.parse(text) as SkillWithResearchResponse
}

export async function completeSession(
  skillId: string,
  body: SessionCompleteBody,
): Promise<SessionCompleteResponse> {
  const res = await fetch(
    `${apiBase}/api/skills/${encodeURIComponent(skillId)}/complete-session`,
    {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    },
  )
  const text = await res.text()
  if (!res.ok) {
    throw new Error(errorFromResponse(res, text))
  }
  return JSON.parse(text) as SessionCompleteResponse
}

File: backend/app/schemas/skills.py
"""API shapes for skills, research, and progress (decoupled from ORM rows)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class SkillCreate(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    notes: Optional[str] = Field(default=None, max_length=20000)


class SkillCreateWithResearch(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    goal: str = Field(min_length=1, max_length=20000)
    level: str = Field(min_length=1, max_length=64, description="e.g. Beginner")
    category: Optional[str] = Field(
        default=None,
        max_length=64,
        description="Preset from skill select, e.g. cooking, music",
    )


class SkillUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=500)
    notes: Optional[str] = Field(default=None, max_length=20000)


class SkillOut(BaseModel):
    id: str
    title: str
    notes: Optional[str]
    context: Optional[dict[str, Any]] = None
    stats_sessions: int = 0
    stats_practice_seconds: int = 0
    stats_level: int = 1
    stats_progress_percent: float = 0.0
    stats_mastered: int = 0
    stats_day_streak: int = 0
    last_practice_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class SessionCompleteBody(BaseModel):
    duration_seconds: int = Field(ge=0, le=86400)
    session_notes: Optional[str] = Field(default=None, max_length=8000)


class SkillSessionSummaryOut(BaseModel):
    id: str
    skill_id: str
    session_number: int
    duration_seconds: int
    summary_text: str
    coach_note: Optional[str]
    progress_delta: float
    level_ups: int
    mastered_delta: int
    input_notes: Optional[str]
    extra: Optional[dict[str, Any]] = None
    created_at: datetime


class SessionCompleteResponse(BaseModel):
    skill: SkillOut
    coach_note: str
    progress_delta: float
    level_ups: int
    mastered_delta: int
    session_summary: SkillSessionSummaryOut
    docs_export_url: Optional[str] = None


class ResearchCreate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=500)
    content: str = Field(min_length=1)
    extra: Optional[dict[str, Any]] = None


class ResearchOut(BaseModel):
    id: str
    skill_id: str
    title: Optional[str]
    content: str
    extra: Optional[dict[str, Any]]
    created_at: datetime


class ProgressCreate(BaseModel):
    kind: str = Field(min_length=1, max_length=64)
    label: Optional[str] = Field(default=None, max_length=500)
    detail: Optional[dict[str, Any]] = None
    metric_value: Optional[float] = None


class ProgressOut(BaseModel):
    id: str
    skill_id: str
    kind: str
    label: Optional[str]
    detail: Optional[dict[str, Any]]
    metric_value: Optional[float]
    created_at: datetime


class SkillWithResearchResponse(BaseModel):
    skill: SkillOut
    research: ResearchOut


class LiveCoachContextOut(BaseModel):
    """Skill + latest research + recent progress for Gemini Live system instructions."""

    skill: SkillOut
    research: Optional[ResearchOut] = None
    progress_events: list[ProgressOut] = Field(default_factory=list)


class LiveSystemInstructionOut(BaseModel):
    system_instruction: str
    live_context_version: str
    skill_id: str
    source_research_id: Optional[str] = None
    source_progress_event_ids: list[str] = Field(default_factory=list)
    truncated: bool = False
